### PR TITLE
overlord/devicestate: support for snap downloads in recovery system handlers

### PR DESCRIPTION
**Pull request by bboozzoo**
During a remodel, some snaps may not be installed yet and are being fetched in
the download tasks. Support fetching the snaps by inspecting the task state of
the downloads.

### CODE REVIEW

File: overlord/devicestate/handlers_systems.go [hunk 1]
```diff
@@ -157,10 +159,40 @@ func (m *DeviceManager) doCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) (err
...
+			taskWithSnapSetup := st.Task(tskID)
+			snapsup, err := snapstate.TaskSnapSetup(taskWithSnapSetup)
+			if err != nil {
+				return nil, false, err
+			}
+			if snapsup.SnapName() != name {
+				continue
+			}
+			// by the time this task runs, the file has already been
+			// downloaded
```
Comment (pedronis) [inline overlord/devicestate/handlers_systems.go:183]: and validated  (we split the phases after validate-snap I think),  which means the open/read below are not risky

File: overlord/devicestate/devicestate_systems_test.go [hunk 2]
```diff
@@ -1236,6 +1240,255 @@ func (s *deviceMgrSystemsCreateSuite) TestDeviceManagerCreateRecoverySystemHappy
...
+	})
+	chg.Set("new-model", string(asserts.Encode(newModel)))
+
+	s.mockStandardSnapsModeenvAndBootloaderState(c)
+
+	s.state.Unlock()
+	s.settle(c)
+	s.state.Lock()
+
+	c.Assert(chg.Err(), ErrorMatches, `(?s).*cannot create a recovery system.*internal error: non-essential but "required" snap "bar" not present.`)
```
Comment (pedronis) [inline overlord/devicestate/devicestate_systems_test.go:1475]: I didn't notice when looking at the original code,  the quotes around required are a bit strange

**Review by pedronis — APPROVED** [review]
thanks, couple small comments

File: overlord/devicestate/devicestate_systems_test.go (hunk 2 above)
Comment (mardy) [reply overlord/devicestate/devicestate_systems_test.go:1475]: Are they? http://www.unnecessaryquotes.com/ ;-)

File: overlord/devicestate/handlers_systems.go [hunk 3]
```diff
@@ -157,10 +159,40 @@ func (m *DeviceManager) doCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) (err
...
+			// created from snaps that are already installed
+			return nil, false, nil
+		}
+
+		// in a remodel scenario, the snaps may need to be fetched, and
+		// thus we can pull the relevant information from the tasks
+		// carrying snap-setup
+
+		for _, tskID := range setup.SnapSetupTasks {
+			taskWithSnapSetup := st.Task(tskID)
```
Comment (stolowski) [inline overlord/devicestate/handlers_systems.go:174]: I guess the tasks are part of the current change being executed so there is no chance for some of them to be gone already?

**Review by stolowski — APPROVED** [review]
LGTM, two tiny questions, and I agree with the comment re "required" in quotes. Thank you

File: overlord/devicestate/handlers_systems.go (hunk 3 above)
Comment (bboozzoo) [reply overlord/devicestate/handlers_systems.go:174]: Yes, the download tasks are part of the current change.

**Comment by bboozzoo** [comment]
The failures are unrelated, @mvo5 @pedronis can you land this PR?

**Pull request merged**